### PR TITLE
[BEAM-12823] TestStream Support in Samza Portable Runner

### PR DESCRIPTION
**Pull request by kw2542**
TestStream Support in Samza Portable Runner

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

**Comment by kw2542** [comment]
Run Portable_Python PreCommit

**Comment by kw2542** [comment]
Run Python_PVR_Flink PreCommit

**Comment by kw2542** [comment]
Run Python_PVR_Flink PreCommit

**Comment by kw2542** [comment]
Run Python_PVR_Flink PreCommit

**Comment by kw2542** [comment]
Run Portable_Python PreCommit

**Comment by kw2542** [comment]
This PR does not change python precommit at all, and the failed tests failed with empty commit as well: https://github.com/apache/beam/pull/15433

**Comment by kw2542** [comment]
Run Python Samza ValidatesRunner

**Comment by kw2542** [comment]
Run XVR_Samza PostCommit

**Comment by kw2542** [comment]
@xinyuiscool @Zhangyx39 Could you help take a look ?

**Comment by kw2542** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by kw2542** [comment]
Run Direct ValidatesRunner

**Comment by kw2542** [comment]
Samza CrossLanguageValidatesRunner Tests & Python Samza ValidatesRunner Tests are currently failing on master as well:

- https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/
- https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/

**Comment by kw2542** [comment]
Run Python Samza ValidatesRunner

**Comment by kw2542** [comment]
Run Portable_Python PreCommit

**Comment by kw2542** [comment]
Run Python_PVR_Flink PreCommit

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java [hunk 1]
```diff
@@ -90,11 +95,53 @@ public void translate(
...
+            return TestStreamTranslation.testStreamFromProtoPayload(payload, coder);
+          } catch (IOException e) {
+            throw new SamzaException("Could not decode TestStream.", e);
+          }
+        };
+
+    final Map<String, String> systemConfig =
+        ImmutableMap.of(
+            "encodedTestStream",
+            Base64Serializer.serializeUnchecked(bytes),
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:135]: Is this being used? Seems the decoder function is already serialized with payload and actual coder, so not sure whether we need to pass in it here.

**Review by xinyuiscool — COMMENTED** [review]
Looks good. One comment about the serialization of the payloads.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java [hunk 2]
```diff
@@ -90,11 +95,53 @@ public void translate(
...
+          (Coder<T>)
+              RehydratedComponents.forComponents(pipeline.getComponents())
+                  .getCoder(payload.getCoderId());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    // the decoder for encodedTestStream
+    SerializableFunction<String, TestStream<?>> testStreamDecoder =
+        string -> {
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:124]: string seems not used?

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java (hunk 1 above)
Comment (kw2542) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:138]: Updated the decoder function not to encode the input anymore.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java (hunk 2 above)
Comment (kw2542) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:120]: Updated

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java [hunk 3]
```diff
@@ -90,11 +97,54 @@ public void translate(
...
 
+  @SuppressWarnings("unchecked")
   @Override
   public void translatePortable(
       PipelineNode.PTransformNode transform,
       QueryablePipeline pipeline,
       PortableTranslationContext ctx) {
-    throw new SamzaException("TestStream is not supported in portable by Samza runner");
+    final ByteString bytes = transform.getTransform().getSpec().getPayload();
+    final Coder<T> coder;
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:107]: Shall we wrap line 107 - 129 into a private static method, e.g. createTestStreamDecoder()? The good thing about static is that your lambda function instance won't hold on to the outer class instance.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java [hunk 4]
```diff
@@ -90,11 +97,54 @@ public void translate(
...
+
+    try {
+      coder =
+          (Coder<T>)
+              RehydratedComponents.forComponents(pipeline.getComponents())
+                  .getCoder(RunnerApi.TestStreamPayload.parseFrom(bytes).getCoderId());
+    } catch (IOException e) {
+      throw new SamzaException(e);
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:115]: Let's throw a RuntimeException() here.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java [hunk 5]
```diff
@@ -90,11 +97,54 @@ public void translate(
...
+    }
+
+    // the decoder for encodedTestStream
+    SerializableFunction<String, TestStream<?>> testStreamDecoder =
+        string -> {
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaTestStreamTranslator.java:120]: Let's have a meaningful name for string. Thx!

**Review by xinyuiscool — APPROVED** [review]
LGTM! Thanks a lot for cleaning up the code.

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

**Comment by kw2542** [comment]
Run Python Samza ValidatesRunner

**Comment by kw2542** [comment]
Run Samza ValidatesRunner

**Comment by kw2542** [comment]
Run Java PreCommit

**Pull request merged**